ocaml-lsp-server/src/custom_requests/req_type_enclosing.ml

```diff
@@ -0,0 +1,162 @@
+       the we do not provide rangeEnd parameter. *)
+    None
+
+let raise_invalid_params ?data ~message () =
+  let open Jsonrpc.Response.Error in
+  raise @@ make ?data ~code:Code.InvalidParams ~message ()
+
+let of_yojson = function
+  | `Assoc params as full_params ->
+    let verbosity = verbosity_of_yojson params in
+    let range_end = range_end_of_yojson params in
+    let open Option.O in
+    let* index = index_of_yojson params in
+    let text_document_position =
+      Lsp.Types.TextDocumentPositionParams.t_of_yojson full_params
+    in
+    Some { index; range_end; verbosity; text_document_position }
+  | _ -> None
+
+let of_yojson_exn = function
+  | None -> raise_invalid_params ~message:"Expected params but received none" ()
+  | Some params -> (
+    match of_yojson params with
+    | Some params -> params
+    | None ->
+      let data =
+        `Assoc
+          [ ("expectedParams", expected_params)
+          ; ("receivedParams", (params :> Json.t))
+          ]
+      in
+      raise_invalid_params ~data ~message:"Unexpected params format" ())
+
+let overlap_with_range_end range = function
+  | None -> true
+  | Some position ->
+    let lend = range.Range.end_ in
+    if lend.line = position.Position.line then
+      lend.character > position.character
+    else lend.line > position.line
+
+let render_result index result =
+  let typ, enclosings =
+    match result with
+    | None -> ("<no information>", [])
+    | Some (typ, enclosings) ->
+      (typ, List.map ~f:Lsp.Types.Range.yojson_of_t enclosings)
+  in
+  `Assoc
+    [ ("index", `Int index)
+    ; ("enclosings", `List enclosings)
+    ; ("type", `String typ)
+    ]
+
+let config_with_given_verbosity config verbosity =
+  let open Mconfig in
+  { config with query = { config.query with verbosity } }
+
+let with_pipeline state uri verbosity with_pipeline =
+  let doc = Document_store.get state.State.store uri in
+  match Document.kind doc with
+  | `Other -> Fiber.return None
+  | `Merlin merlin ->
+    let open Fiber.O in
+    let* config = Document.Merlin.mconfig merlin in
+    Document.Merlin.with_configurable_pipeline_exn
+      ~config:(config_with_given_verbosity config verbosity)
+      merlin
+      with_pipeline
+
+let make_enclosing_command position index =
+  Query_protocol.Type_enclosing (None, position, Some index)
+
+let get_logical_position tdp =
+  let p = tdp.Lsp.Types.TextDocumentPositionParams.position in
+  Position.logical p
+
+let get_first_enclosing_index range_end enclosings =
+  let rec aux i = function
+    | (loc, _, _) :: xs ->
+      let range = Range.of_loc loc in
+      if overlap_with_range_end range range_end then Some i else aux (succ i) xs
+    | _ -> None
+  in
+  aux 0 enclosings
+
+let dispatch_type_enclosing text_document_position index range_end pipeline =
+  let position = get_logical_position text_document_position in
+  let dummy_command = make_enclosing_command position (-1) in
```

Comment:
We should comment somewhere about this hack.

Reply:
And problably only rely on it when renage_end <> None

Reply:
Fixed on the last commit